internal/controller/management/projects/projects.go

```diff
@@ -441,6 +452,87 @@ func (r *reconciler) ensureAPIAdminPermissions(
 	return nil
 }
 
+func (r *reconciler) ensureControllerPermissions(
+	ctx context.Context,
+	project *kargoapi.Project,
+) error {
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"project", project.Name,
+		"namespace", project.Name,
+	)
+
+	// Get all ServiceAccounts labeled as controller ServiceAccounts
+	controllerSAs := &corev1.ServiceAccountList{}
+	if err := r.client.List(
+		ctx, controllerSAs,
+		client.InNamespace(r.cfg.KargoNamespace),
+		client.MatchingLabels{
+			r.cfg.ControllerServiceAccountLabelKey: r.cfg.ControllerServiceAccountLabelValue,
+		},
+	); err != nil {
+		return fmt.Errorf("error listing controller ServiceAccounts: %w", err)
+	}
+
+	// Create/update a RoleBinding for each ServiceAccount
+	for _, controllerSA := range controllerSAs.Items {
+		sa := &controllerSA
+		if controllerutil.AddFinalizer(sa, kargoapi.FinalizerName) {
+			if err := r.client.Update(ctx, sa); err != nil {
+				return fmt.Errorf(
+					"error adding finalizer to controller ServiceAccount %q in namespace %q: %w",
+					sa.Name, sa.Namespace, err,
+				)
+			}
+		}
+
+		roleBindingName := getRoleBindingName(sa.Name)
+		saLogger := logger.WithValues(
+			"serviceAccount", sa.Name,
+			"serviceAccount.namespace", sa.Namespace,
```

Comment:
Given it's structured logging, I wonder if we shouldn't make this structured. I.e.:

```suggestion
			"serviceAccount": types.NamespacedName{
				Name: sa.Name,
				Namespace: sa.Namespace,
			},
```

Reply:
I'd been consciously avoiding that until now just out of a innate sense that it would make log messages much uglier to look at. Let me try it and see how I feel about the result.

Reply:
I tried and it's worse than I thought.

This:

```
logger := logging.LoggerFromContext(ctx).WithValues("serviceAccount", req.NamespacedName)
logger.Debug("reconciling ServiceAccount")
```

Yields messages like this:

```
time="2024-10-17T14:44:01Z" level=debug msg="reconciling ServiceAccount" serviceAccount="{kargo-controller kargo}"
```

Notice that values are shown, but keys are not.

This may be a specific quirk of the underlying logger implementation, but until we change that, I'm not sure that this is an improvement.

wdyt?

Reply:
I forgot we do not have JSON logging enabled by default. Given this, the current option does indeed seem better.

Reply:
Definitely something to improve soon-ish.